constantine/math_compiler/impl_curves_ops_jacobian.nim

```diff
@@ -0,0 +1,663 @@
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = cd.name & "_neg_impl"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([a]),
+          {kHot}):
+    tagParameter(1, "sret")
+
+    let ai = llvmParams
+    let aEc = asy.asEcPointJac(ai, cd.curveTy)
+    asy.neg(cd.fd, aEc.getY().buf, aEc.getY().buf)
+
+    asy.br.retVoid()
+
+  asy.callFn(name, [a])
+
+proc cneg*(asy: Assembler_LLVM, cd: CurveDescriptor, a, c: ValueRef) {.used.} =
+  ## Generate an internal elliptic curve point conditional negation proc
+  ## with signature
+  ##   `void name(CurveType a, bool condition)`
+  ## with `a` the EC curve point to be negated if `condition` is true.
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = cd.name & "_cneg_impl"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([a, c]),
+          {kHot}):
+    tagParameter(1, "sret")
+
+    let (ai, ci) = llvmParams
+    let aEc = asy.asEcPointJac(ai, cd.curveTy)
+    asy.cneg(cd.fd, aEc.getY().buf, aEc.getY().buf, ci)
+
+    asy.br.retVoid()
+
+  asy.callFn(name, [a, c])
+
+proc sum*(asy: Assembler_LLVM, cd: CurveDescriptor, r, p, q: ValueRef) =
+  ## Generate an internal elliptic curve point addition proc
+  ## with signature
+  ##   `void name(CurveType r, CurveType p, CurveType q)`
+  ## with `a` and `b` EC curve point elements to be added.
+  ## The result is stored in `r`.
+  ##
+  ## Generates a call, so that we one can use this proc as part of another procedure.
+  let name = cd.name & "_sum_impl"
+  asy.llvmInternalFnDef(
+          name, SectionName,
+          asy.void_t, toTypes([r, p, q]),
+          {kHot}):
+    tagParameter(1, "sret")
+    let (ri, pi, qi) = llvmParams
+    let Q = asy.asEcPointJac(qi, cd.curveTy)
+    let P = asy.asEcPointJac(pi, cd.curveTy)
+    let rA = asy.asEcPointJac(ri, cd.curveTy)
+
+    ## Helper templates to allow the logic below to be roughly equivalent to the regular
+    ## CPU code in `ec_shortweierstrass_jacobian.nim`.
+
+    # Make finite field point operations nicer
+    fieldOps(asy, cd.fd)
+    # And EC points
+    ellipticOps(asy, cd)
+
+    var
+      Z1Z1 = asy.newField(cd.fd)
+      U1   = asy.newField(cd.fd)
+      S1   = asy.newField(cd.fd)
+      H    = asy.newField(cd.fd)
+      R    = asy.newField(cd.fd)
+
+    block: # Addition-only, check for exceptional cases
+      var
+        Z2Z2 = asy.newField(cd.fd)
+        U2   = asy.newField(cd.fd)
+        S2   = asy.newField(cd.fd)
+
+      ## XXX: Similarly to `prod` below, `skipFinalSub = true` here, also breaks the
+      ## code at a different point.
+      Z2Z2.square(Q.z, skipFinalSub = false)
+      ## XXX: If we set this `skipFinalSub` to true, the code will fail to produce the
+      ## correct result in some cases. Not sure yet why. See `tests/gpu/t_ec_sum.nim`.
+      S1.prod(Q.z, Z2Z2, skipFinalSub = false)
```

Comment:
Either of these cause problems in the code for multiple iterations in the mentioned test case. The two below (Z1Z1 and S2) work fine though. _Maybe_ we want to avoid it everywhere for the time being out of an abundance of caution though.